assert: deprecate assert.fail partially

BridgeAR opened the pull request:
Using `assert.fail()` with more than one argument is not intuitive
to use and has no benefit over using a message on its own.

Therefore this introduces a runtime deprecation in case it is used
in that way.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
assert

lpinca commented on doc/api/assert.md:
```diff
@@ -406,27 +406,52 @@ parameter is an instance of an [`Error`][] then it will be thrown instead of the
-[`Error.captureStackTrace`]). If no arguments are given, the default message
-`Failed` will be used.
+* `actual` {any} _Deprecated if not used as message (see below)_
+* `expected` {any}_Deprecated (see below)_
```
Nit: space after first `_`.

lpinca commented on lib/assert.js:
```diff
@@ -80,8 +82,18 @@ function fail(actual, expected, message, operator, stackStartFn) {
+      warned = true;
+      process.emitWarning(
+        'assert.fail() with more than one argument is deprecated. ' +
+          'Please use assert.strictEqual instead or only pass a message.',
```
Nit: `assert.strictEqual()` for consistency.

vsemozhetbyt commented on doc/api/deprecations.md:
```diff
@@ -818,6 +818,15 @@ a future version at which point only authentication tag lengths of 128, 120,
+
+Type: Runtime
+
+Using `assert.fail()` with more than one argument has no benefit over writing a
```
Nit: a individual -> an individual.

BridgeAR commented:
Comments addressed.

CI https://ci.nodejs.org/job/node-test-pull-request/12786/

joyeecheung commented on doc/api/assert.md:
```diff
@@ -406,27 +406,52 @@ parameter is an instance of an [`Error`][] then it will be thrown instead of the
 ## assert.fail(actual, expected[, message[, operator[, stackStartFunction]]])
```
Also, that way we can put a `deprecated` entry in the YAML so the whole thing would be labeled with ``deprecated` in the TOC.

BridgeAR replied:
I agree that it looks confusing. We just never had such a case before but I agree that this is probably the best way to solve this. I will change it shortly.

BridgeAR replied:
Done

BridgeAR commented:
New CI https://ci.nodejs.org/job/node-test-pull-request/12805/

vsemozhetbyt commented on doc/api/assert.md:
```diff
@@ -403,30 +403,64 @@ parameter is an instance of an [`Error`][] then it will be thrown instead of the
 * `expected` {any}
-* `message` {any} **Default:** `'Failed'`
-* `operator` {string} **Default:** '!='
+* `operator` {string} **Default:** '!='_
```
Unnecessary underscore?

jasnell commented on doc/api/assert.md:
````diff
@@ -440,21 +465,11 @@ assert.fail(1, 2, new TypeError('need array'));
 ```
 
-*Note*: In the last two cases `actual`, `expected`, and `operator` have no
+*Note*: In the last three cases `actual`, `expected`, and `operator` have no
````
Perhaps drop the `*:Note*: ` part of this now? It's not super necessary.

jasnell commented on doc/api/deprecations.md:
```diff
@@ -818,6 +818,15 @@ a future version at which point only authentication tag lengths of 128, 120,
 [NIST SP 800-38D][].
 
+<a id="DEP0XXX"></a>
+### DEP0XXX: Using `assert.fail()` with more than one argument.
```
s/DEP0XXX/DEP00XX

joyeecheung commented on test/parallel/test-assert-fail.js:
```diff
@@ -5,6 +5,12 @@
 const common = require('../common');
 const assert = require('assert');
 
+common.expectWarning(
```
Can you put the tests with the valid signatures to another file?

BridgeAR replied:
Done

BridgeAR commented:
Rebased due to conflicts and while doing so I also rewrote `common.expectsError` to `assert.throws` and found a mistake in one of the old tests because of that and fixed it. I also addressed @joyeecheung comment to split the tests into two files.

BridgeAR commented:
New CI https://ci.nodejs.org/job/node-test-pull-request/12909/

BridgeAR commented:
Landed in 70dcacd7101be14321b8a1c05de75a78b4656704
